Sources/SwiftProtobuf/TextDecoder.swift

```diff
@@ -0,0 +1,211 @@
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
+///
+/// Test format decoding engine.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+import Swift
+
+///
+/// Provides a higher-level interface to the token stream coming
+/// from a TextScanner.  In particular, this provides single-token
+/// pushback and convenience functions for iterating over complex
+/// structures.
+///
+public struct TextDecoder {
+    private var scanner: TextScanner
+    public var complete: Bool {return scanner.complete}
+
+    internal init(text: String, extensions: ExtensionSet? = nil) {
+        scanner = TextScanner(text: text, tokens: [], extensions: extensions)
+    }
+
+    internal init(scanner: TextScanner) {
+        self.scanner = scanner
+    }
+
+    internal mutating func decodeFullObject<M: Message>(message: inout M, terminator: TextToken?) throws {
+        guard let nameProviding = (M.self as? ProtoNameProviding.Type) else {
+            throw DecodingError.missingFieldNames
+        }
+        while let token = try scanner.nextKey() {
+            switch token {
+            case .identifier(let key):
+                let protoFieldNumber: Int
+                if key.hasPrefix("[") {
+                    // Extension key
+                    if let n = scanner.extensions?.fieldNumberForProto(messageType: M.self, protoFieldName: key) {
+                        protoFieldNumber = n
+                    } else {
+                        throw DecodingError.unknownField
+                    }
+                } else {
+                    // Regular key; look it up on the message
+                    if let n = nameProviding._protobuf_fieldNames.fieldNumber(forProtoName: key) {
+                        protoFieldNumber = n
+                    } else {
+                        throw DecodingError.unknownField
+                    }
+                }
+                var fieldDecoder: FieldDecoder = TextFieldDecoder(scanner: scanner)
+                try message.decodeField(setter: &fieldDecoder, protoFieldNumber: protoFieldNumber)
+            default:
+                if terminator != nil && terminator == token {
+                    return
+                }
+                throw DecodingError.malformedText
+            }
+            try scanner.skipOptionalSeparator()
+        }
+        if terminator == nil {
+            return
+        } else {
+            throw DecodingError.truncatedInput
+        }
+    }
+}
+
+struct TextFieldDecoder: FieldDecoder {
+    var rejectConflictingOneof: Bool {return true}
```

Comment:
I believe this is optional in the C++ implementation.  I think the C++ version has way too many options, so I've tried to implement this first with no options at all, trying to find the most generally useful approach.  I figure it's easy to introduce options later as real need arises.